src/dvi.rs

```diff
@@ -102,39 +116,127 @@ impl<Channels: DmaChannelList> DviInst<Channels> {
             let buf_ix = (y as usize / VERTICAL_REPEAT) % N_TMDS_BUFFERS;
-            let stride = self.timing.horizontal_words() as usize * N_CHANNELS * buf_ix;
-            let buf = unsafe { self.tmds_buf.as_ptr().add(stride) };
-            let channel_stride = if N_CHANNELS == 1 {
-                0
-            } else {
-                self.timing.horizontal_words()
-            };
-            self.dma_list_active.update_scanline(buf, channel_stride);
+            if self.available[buf_ix] {
+                let stride = self.timing.horizontal_words() as usize * N_CHANNELS * buf_ix;
+                let buf = unsafe { self.tmds_buf.as_ptr().add(stride) };
+                let channel_stride = if N_CHANNELS == 1 {
+                    0
+                } else {
+                    self.timing.horizontal_words()
+                };
+                self.dma_list_active.update_scanline(buf, channel_stride);
+                return true;
+            }
         }
+        false
     }
 
+    /// Render a scanline into a TMDS buffer.
+    ///
+    /// This function is called even if the corresponding line buffer is not
+    /// available, so the display list can be advanced.
     #[link_section = ".data"]
-    fn render(&mut self) {
-        if let Some(y) = self
-            .timing_state
-            .v_scanline_index(&self.timing, TMDS_PIPELINE_SLACK)
-        {
+    fn render(&mut self, y: u32, available: bool) {
+        let buf_ix = y as usize % N_TMDS_BUFFERS;
+        let line_size = self.timing.horizontal_words() as usize * N_CHANNELS;
+        let line_start = line_size * buf_ix;
+        let tmds_slice = &mut self.tmds_buf[line_start..][..line_size];
+        self.scan_render.render_scanline(tmds_slice, y, available);
+    }
+
+    /// Schedule the rendering of a line buffer.
+    ///
+    /// The line buffers are rendered outside the main interrupt handler,
+    /// striped across both cores.
+    #[link_section = ".data"]
+    fn schedule_line_render(&mut self) {
+        let offset = TMDS_PIPELINE_SLACK + (N_LINE_BUFS * VERTICAL_REPEAT) as u32;
+        if let Some(y) = self.timing_state.v_scanline_index(&self.timing, offset) {
             if y % VERTICAL_REPEAT as u32 == 0 {
                 let y = y / VERTICAL_REPEAT as u32;
-                let buf_ix = y as usize % N_TMDS_BUFFERS;
-                let line_size = self.timing.horizontal_words() as usize * N_CHANNELS;
-                let line_start = line_size * buf_ix;
-                let tmds_slice = &mut self.tmds_buf[line_start..][..line_size];
-                self.scan_render.render_scanline(tmds_slice, y);
+                self.scan_render.schedule_line_render(y);
             }
         }
     }
 }
 
+/// We dedicate core 1 to running the primary video interrupt and also
+/// background rendering tasks.
+#[link_section = ".data"]
+pub fn core1_main<PIO, SliceId, Pos, Neg, RedPos, RedNeg, GreenPos, GreenNeg, BluePos, BlueNeg>(
```

Comment:
These generics really are something..... WE could maybe do something like in #9 where we use a marker trait for all the pin generics. The reason I didn't do this before was that you no longer get any type inference as you have to make a type to implement the marker trait on, but then anywhere you use these 6/8 pins you just can refer to the marker trait rather than all pins individually.